packages/api-server/api_server/rmf_io/health_watchdog.py

```diff
@@ -24,41 +31,239 @@ def __init__(
             ),
-        ).subscribe(report_door_health, scheduler=self.scheduler)
+        )
+
+    def _watch_door_health(self, building_map):
+        def door_mode_to_health(data: Tuple[str, DoorState]):
+            state = data[1]
+
+            if state is None:
+                return models.DoorHealth(
+                    id_=data[0],
+                    health_status=models.HealthStatus.UNHEALTHY,
+                    health_message="unknown",
+                )
+            if state.current_mode.value in (
+                DoorMode.MODE_CLOSED,
+                DoorMode.MODE_MOVING,
+                DoorMode.MODE_OPEN,
+            ):
+                return models.DoorHealth(
+                    id_=state.door_name,
+                    health_status=models.HealthStatus.HEALTHY,
+                )
+            if state.current_mode.value == DoorMode.MODE_OFFLINE:
+                return models.DoorHealth(
+                    id_=state.door_name,
+                    health_status=models.HealthStatus.UNHEALTHY,
+                    health_message="offline",
+                )
+            return models.DoorHealth(
+                id_=state.door_name,
+                health_status=models.HealthStatus.UNHEALTHY,
+                health_message="unknown",
+            )
+
+        doors: List[Door] = []
+        if building_map:
+            for level in building_map.levels:
+                level: Level
+                for door in level.doors:
+                    doors.append(door)
+
+        def to_door_health(data: Tuple[str, bool]):
+            id_ = data[0]
+            has_heartbeat = data[1]
+            if has_heartbeat:
+                return models.DoorHealth(
+                    id_=id_, health_status=models.HealthStatus.HEALTHY
+                )
+            return models.DoorHealth(
+                id_=id_,
+                health_status=models.HealthStatus.DEAD,
+                health_message="heartbeat failed",
+            )
+
+        keys = [x.name for x in doors]
+        initial_values: Sequence[Tuple[str, Any]] = [(k, None) for k in keys]
+        obs = rx.merge(
+            rx.of(*initial_values),
+            self.rmf.door_states.pipe(op.map(lambda x: (x.door_name, x))),
+        )
+
+        door_mode_health = obs.pipe(
+            op.map(door_mode_to_health),
+            op.timestamp(),
+        )
+
+        heartbeat_health = obs.pipe(
+            self._watch_heartbeat(lambda x: x[0]),
+            op.map(to_door_health),
+            op.timestamp(),
+        )
+
+        sub = heartbeat_health.pipe(
+            op.combine_latest(door_mode_health),
+            most_critical(),
+        ).subscribe(self._report_health(self.rmf.door_health), scheduler=self.scheduler)
+        self.watchers.append(sub)
+
+    def _watch_lift_health(self, building_map):
+        lifts: Sequence[Lift] = building_map.lifts if building_map else []
+
+        def to_lift_health(data: Tuple[str, bool]):
```

Comment:
I think that if we follow what you mentioned

> Should we make the distinction of health status based on its ability to response to rmf

The only thing that it's missing (i think) is to handle the     `static readonly MODE_OFFLINE = 4;` state of the lift